cla_backend/apps/reports/tests/test_mi_sla_report.py

```diff
@@ -29,43 +35,252 @@ def patch_field(cls, field_name, dt):
+        case = self.make_case(dt)
+        user = make_user()
+        make_recipe("call_centre.operator", user=user)
+
+        # Create CB 1
+        self.schedule_callback(case, user, dt)
+
+        date_range = (dt - datetime.timedelta(days=2), dt + datetime.timedelta(days=2))
+        values = self.get_report(date_range)
+
+        self.assertFalse(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_cb1_uncalled_and_current_time_after_sla1(self):
+        # SLA 1 miss is TRUE - No action within sla1 window
+        # SLA 2 miss is FALSE - Current time is within sla2 window
+        dt = _make_datetime(hour=9, minute=1) - datetime.timedelta(days=1)
+        case = self.make_case(dt)
+        user = make_user()
+        make_recipe("call_centre.operator", user=user)
+
+        # Create CB 1
+        self.schedule_callback(case, user, dt)
+
+        date_range = (dt - datetime.timedelta(days=2), dt + datetime.timedelta(days=2))
+        values = self.get_report(date_range)
+
+        self.assertTrue(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_cb2_scheduled_WITHIN_sla1_window_and_current_time_WITHIN_sla2_window(self):
+        # SLA 1 miss is FALSE - CB2 Scheduled within sla1 window
+        # SLA 2 miss is FALSE - Current time is within sla2 window
+        dt = _make_datetime(hour=9, minute=1) - datetime.timedelta(days=1)
+        case = self.make_case(dt)
+        user = make_user()
+        make_recipe("call_centre.operator", user=user)
+
+        # Create CB 1
+        self.schedule_callback(case, user, dt)
+        # Create CB2
+        self.schedule_callback(case, user, dt + datetime.timedelta(minutes=25))
+
+        date_range = (dt - datetime.timedelta(days=2), dt + datetime.timedelta(days=1))
+        values = self.get_report(date_range)
+
+        self.assertFalse(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_cb2_scheduled_BEFORE_sla1_window_and_current_time_WITHIN_sla2_window(self):
+        # SLA 1 miss is TRUE - CB2 Scheduled before sla1 window
+        # SLA 2 miss is FALSE - Current time is within sla2 window
+        dt = _make_datetime(hour=9, minute=1) - datetime.timedelta(days=1)
+        case = self.make_case(dt)
+        user = make_user()
+        make_recipe("call_centre.operator", user=user)
+
+        # Create CB 1
+        self.schedule_callback(case, user, dt)
+        # Create CB2
+        self.schedule_callback(case, user, dt + datetime.timedelta(minutes=-5))
+
+        date_range = (dt - datetime.timedelta(days=2), dt + datetime.timedelta(days=1))
+        values = self.get_report(date_range)
+
+        self.assertTrue(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_cb2_scheduled_AFTER_sla1_window_and_current_time_WITHIN_sla2_window(self):
+        # SLA 1 miss is TRUE  - CB2 Scheduled after sla1 window
```

Comment:
I don't understand how we know if this is an SLA1 miss. The time (which is outside the SLA1 window) is the time the CB2 is scheduled for, right? Rather than the time the CB1 was attempted and failed? Am I missing something?

Reply:
I have now tidied this up by renaming the method `test_cb2_created_AFTER_sla1_window_and_current_time_WITHIN_sla2_window` to make it clear that we are testing the CB2 created time rather the time it was scheduled.

I have also changed the comment `# CB2 Scheduled after sla1 window` to `CB2 created after sla1 window`

We only know that they attempted to contact the user if the call was answered or the scheduled another callback

So for the following workflow:
CB1 is scheduled for 10-10:30 AM
CB2 is created at 10:31 AM and scheduled for 11-11:30 AM

In that work they have missed SLA1 one because the CB2 was created after SLA1 window.